drivers/sdcard/sdcard.py

```diff
@@ -159,27 +164,19 @@ def cmd(self, cmd, arg, crc, final=0, release=True):
         self.spi.write(b'\xff')
         return -1
 
-    def cmd_nodata(self, cmd):
-        self.spi.write(cmd)
```

Comment:
This is actually an invalid command: it lacks the command payload like in `cmd()`. It's so similar to `cmd()` anyway that I've merged it into that function.

Reply:
Ok.  But it seems the way it returns "success" has changed: in cmd_nodata() the byte read must be 0xff, and then it returns 0.  But in cmd() the final byte read must have the top bit (0x80) cleared, and then it returns the whole response byte.  Was cmd_nodata() really that broken that this old logic of waiting for 0xff was wrong?

Reply:
tl;dr: I think the old logic was broken.

To be honest, I don't know. I mostly based myself on this resource:
https://luckyresistor.me/cat-protector/software/sdcard-3/ (see "stopping the read process")
The old logic didn't work and the current one does for two different cards (one very old 256MB Kingston card and one 16GB card I think I got with the Raspberry Pi). I *think* the old logic was just really broken. Note that with the old logic, the first read returns success but all later reads only return 0xff.
My best guess: the previously tested card implicitly ended the read process with the CS going high. Other cards work a bit different internally. And nobody has noticed this problem as most people will only read up to 512 bytes at once.

Another useful resource: http://elm-chan.org/docs/mmc/mmc_e.html